gcore/src/utils.rs

```diff
@@ -50,16 +51,51 @@ pub mod ext {
     ///
     /// #[no_mangle]
     /// extern "C" fn handle() {
-    ///     ext::debug("Hello, world!").expect("Unable to log");
+    ///     ext::debug("Hello, world!");
     /// }
     /// ```
     #[cfg(any(feature = "debug", debug_assertions))]
-    pub fn debug(data: &str) -> Result<()> {
-        let data_len = data.len().try_into().map_err(|_| Error::SyscallUsage)?;
+    pub fn debug(data: &str) {
+        unsafe { gsys::gr_debug(data.as_ptr(), data.len() as u32) }
+    }
+
+    /// Same as [`debug`] but uses a stack-allocated buffer.
+    ///
+    /// Note: message size is limited to
+    /// [`MAX_BUFFER_SIZE`](crate::stack_buffer::MAX_BUFFER_SIZE).
+    #[cfg(any(feature = "debug", debug_assertions))]
+    pub fn stack_debug(args: fmt::Arguments<'_>) {
+        struct StackFmtWriter<'a> {
+            buf: &'a mut [MaybeUninit<u8>],
+            pos: usize,
+        }
+
+        impl fmt::Write for StackFmtWriter<'_> {
+            fn write_str(&mut self, s: &str) -> fmt::Result {
+                let upper_bound = (self.pos + s.len()).min(MAX_BUFFER_SIZE);
+                if let Some(buf) = self.buf.get_mut(self.pos..upper_bound) {
+                    let buf = buf as *mut [MaybeUninit<u8>] as *mut [u8];
+                    let s = &s.as_bytes()[..buf.len()];
```

Comment:
we could avoid panicking here and return `fmt::Error`. instead of `[..buf.len()]` use `s.as_bytes().get(..buf.len())`.

Reply:
It's not panicking. It's requirement of `copy_from_slice` to both slices to be equal length and trimming in case of message exceeds buffer

Reply:
`ops::Index` (`[..buf.len()]`) is panicking operation